### PR TITLE
service/header: Verify hash matches header retrieved in `RequestByHash()`

### LINKED ISSUE #219: service/header: Exchange should verify that requested hash matches the header
Currently, we don't verify a received Header against the hash used for requesting. We should fix this security breach.

### PR DESCRIPTION
**Pull request by renaynay**
Resolves #219

TODO 
- [x] tests
- [x] implement same check for p2p ex

### CODE REVIEW

**Review by liamsi — DISMISSED** [review]
Looks correct. Will the TODOs in the opening comment be resolved here or in follow up PRs?

**Review by liamsi — DISMISSED** [review]
image

**Review by adlerjohn — APPROVED** [review]
ahhh I'm approoooving!

File: service/header/core_exchange.go [hunk 1]
```diff
@@ -56,7 +58,19 @@ func (ce *CoreExchange) RequestByHash(ctx context.Context, hash []byte) (*Extend
...
-	return ce.generateExtendedHeaderFromBlock(block)
+	extHeader, err := ce.generateExtendedHeaderFromBlock(block)
+	if err != nil {
+		return nil, err
+	}
+	// verify hashes match
+	if !hashMatch(hash, extHeader.Hash().Bytes()) {
+		return nil, fmt.Errorf("incorrect hash in header: expected %x, got %x", hash, extHeader.Hash().Bytes())
+	}
+	return extHeader, nil
```
Comment (Wondertan) [inline service/header/core_exchange.go:69]: We also need to have a ValidateBasic on it

Comment (Wondertan) [reply service/header/core_exchange.go:69]: So we can see other fields are matching

File: service/header/exchange.go [hunk 2]
```diff
@@ -191,6 +191,9 @@ func (ex *exchange) RequestByHash(ctx context.Context, hash []byte) (*ExtendedHe
 	if err != nil {
 		return nil, err
 	}
+	if !hashMatch(headers[0].Hash().Bytes(), hash) {
+		return nil, fmt.Errorf("incorrect hash in header: expected %x, got %x", hash, headers[0].Hash().Bytes())
```
Comment (Wondertan) [inline service/header/exchange.go:195]:
```suggestion
		return nil, fmt.Errorf("header hash mismatch: expected %x, got %x", hash, headers[0].Hash().Bytes())
```

**Review by liamsi — DISMISSED** [review]
Still approoooving anyways!

**Review by Wondertan — DISMISSED** [review]

**Review by liamsi — APPROVED** [review]

**Review by Wondertan — APPROVED** [review]

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**